src/smart-components/workspaces/managed-selector/WorkspacesStore.tsx

```diff
@@ -0,0 +1,151 @@
+import React, { PropsWithChildren, createContext, useEffect, useMemo, useReducer } from 'react';
+import { TreeViewWorkspaceItem } from './TreeViewWorkspaceItem';
+import Workspace from './Workspace';
+
+interface WorkspacesStore {
+  isWorkspacesMenuExpanded: boolean;
+  isFetchingWorkspacesFromRBAC: boolean;
+  isFetchingWorkspacesFromRBACError: boolean;
+  selectedWorkspace?: TreeViewWorkspaceItem;
+  fetchedWorkspaces: Workspace[];
+  workspaceTree?: TreeViewWorkspaceItem;
+}
+
+const initialStore: WorkspacesStore = {
+  isWorkspacesMenuExpanded: false,
+  isFetchingWorkspacesFromRBAC: false,
+  isFetchingWorkspacesFromRBACError: false,
+  selectedWorkspace: undefined,
+  fetchedWorkspaces: [],
+  workspaceTree: undefined,
+};
+
+const createWorkspacesStore = () => {
+  const store = { ...initialStore };
+
+  const subs = new Map<string, () => void>();
+
+  const notify = () => {
+    subs.forEach((sub) => sub());
+  };
+
+  const subscribe = (callback: () => void) => {
+    const id = crypto.randomUUID();
+    subs.set(id, callback);
+    return () => {
+      subs.delete(id);
+    };
+  };
+
+  const setIsWorkspacesMenuExpanded = (isExpanded: boolean) => {
+    store.isWorkspacesMenuExpanded = isExpanded;
+    notify();
```

Comment:
Instead of calling the notify in each function, you can create decorator function and wrap all calbacks that should trigger the notify.

```ts
const bar = <A extends unknown[], R>(cb: (...args: A) => R) => {
  return (...args: A): R => {
    const cbRes = cb(...args);
    notify();
    return cbRes;
  };
};

const setIsFetchingWorkspacesFromRBAC = foo((isFetching: boolean) => {
  store.isFetchingWorkspacesFromRBAC = isFetching;
})
```